viewing_party/party.py

```diff
@@ -1,23 +1,139 @@
+        "title": title,
+        "genre": genre, 
+        "rating": rating
+        }
+    else:
+        return None
+        
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data 
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    # watchlist = user_data["watchlist"]
+    # watched = user_data["watched"]
+
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+    return user_data
+
+
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    list_of_ratings = []
+    for movie in user_data["watched"]:
+        list_of_ratings.append(movie["rating"])
+    if len(user_data["watched"]) == 0:
+        avg_rating = 0
+    else:
+        avg_rating = sum(list_of_ratings)/len(list_of_ratings)
+    return avg_rating
+
+def get_most_watched_genre(user_data):
+    genre_list = []
+    for movie in user_data["watched"]:
+        genre_list.append(movie["genre"])
+
+    if user_data["watched"] == []:
+        return None
+    else:
+        counter = 0
+        for genre in genre_list:
+            frequency = genre_list.count(genre)
+            if frequency > counter:
+                counter = frequency
+                most_watched_genre = genre
+            return most_watched_genre
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    unique_watched = []
+    friends_movies = set()
+    users_movies = set()
+    # We want to access "friends"
+    # and for every friend in there, we want to 
+    # access a movie in their "watched" list of
+    # movie dictionaries.
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    for movie in user_data["watched"]:
+        users_movies.add(movie["title"])
+    unique_movies = users_movies - friends_movies
+    for movie in user_data["watched"]:
+        if movie["title"] in unique_movies:
+            unique_watched.append(movie)
+    return unique_watched
+
+def get_friends_unique_watched(user_data):
+    unique_watched = []
+    friends_movies = set()
+    users_movies = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    for movie in user_data["watched"]:
+        users_movies.add(movie["title"])
+    unique_movies = friends_movies - users_movies
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in unique_movies:
+                if movie not in unique_watched:

```

Comment:
Since `unique_watched` is a `list`, this `in` check must iterate through the contents to check whether the movie is there. 

Here, we do need to build a list of movie detail records, not just the title. So `unique_movies` must be a list. However, if we still wanted to gain the constant in lookup even for the result list, we could create another set in which we store only the titles of the movies we've added to `unique_movies`, and look up a movie's title in that set before adding it to the result list, and adding the title to the set if we do end up adding a movie to the result list.

As it is, doing the `in` checks (which are linear) in a loop (essentially linear) gives us an overall quadratic time complexity. But we could get this down to linear by building that extra set, since the operations in the loop would become constant.